src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala

```diff
@@ -221,41 +225,39 @@ private class AppTaskLauncherActor(
       log.info("Task launch for '{}' was accepted. {}", op.taskId, status)
   }
 
-  private[this] def receiveTaskStatusUpdate: Receive = {
-    case TaskStatusUpdate(_, taskId, status) =>
-      def handleTask(oldTask: Task) = {
-        oldTask.update(TaskStateOp.MesosUpdate(status, clock.now())) match {
-          case TaskStateChange.Update(updatedTask) =>
-            log.debug("updating status of {}", taskId)
-            tasksMap += taskId -> updatedTask
-
-          case TaskStateChange.Expunge =>
-            log.debug("{} finished", taskId)
-            removeTask(taskId)
+  private[this] def receiveTaskTrackerUpdate: Receive = {
+    case ActorTaskTrackerUpdateSubscriber.HandleTaskStateChange(stateChange) =>
+      stateChange match {
+        case TaskStateChange.Update(updatedTask) =>
+          log.debug("updating status of {}", updatedTask.taskId)
+          tasksMap += updatedTask.taskId -> updatedTask
 
-          case TaskStateChange.NoChange =>
-            log.debug("no change for {}", taskId)
-
-          case TaskStateChange.Failure(cause) =>
-            log.warning("Task state update failed for {}. Reason: {}", taskId, cause.getMessage)
-        }
-      }
-
-      tasksMap.get(taskId) match {
-        case Some(oldTask) => handleTask(oldTask)
-        case _             => log.warning("ignore update of unknown {}", taskId)
-      }
-
-      if (status.terminal) {
-        // If the app has constraints, we need to reconsider offers that
-        // we already rejected. E.g. when a host:unique constraint prevented
-        // us to launch tasks on a particular node before, we need to reconsider offers
-        // of that node after a task on that node has died.
-        if (app.constraints.nonEmpty) {
-          maybeOfferReviver.foreach(_.reviveOffers())
-        }
+        case TaskStateChange.Expunge(taskId) =>
+          log.debug("{} finished", taskId)
+          removeTask(taskId)
+          // If the app has constraints, we need to reconsider offers that
+          // we already rejected. E.g. when a host:unique constraint prevented
+          // us to launch tasks on a particular node before, we need to reconsider offers
+          // of that node after a task on that node has died.
+          if (app.constraints.nonEmpty) {
+            maybeOfferReviver.foreach(_.reviveOffers())
+          }
+
+        case _ =>
+          // FIXME (3221): Only TaskUpdated and TaskRemoved are propagated to the taskTracker
+          // right now, so we'll only receive Update/Expunge here. It would be good to receive
+          // Failures as well along with an actualState: Option[Task] so we can update/remove
+          // the internal tasksMap. besides, it's ugly to only partly receive stateChanges
+          // (although Failure and NoChange are no changes in that sense)
+          log.debug("receiveTaskTrackerUpdate: ignoring unexpected stateChange {}", stateChange)
```

Comment:
I don't see that receiving failures/noops here makes sense. Let's discuss why you think that it does.